optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/common/AbstractRestServiceImpl.java

```diff
@@ -12,9 +11,13 @@
 
     protected void validateTenantIdParameter(Integer tenantId, AbstractPersistable persistable) {
         if (!Objects.equals(persistable.getTenantId(), tenantId)) {
-            throw new IllegalStateException("The tenantId (" + tenantId
-                    + ") does not match the persistable (" + persistable + ")'s tenantId (" + persistable.getTenantId() + ").");
+            throw new IllegalStateException("The tenantId (" + tenantId + ") does not match the persistable (" + persistable + ")'s tenantId (" + persistable.getTenantId() + ").");
         }
     }
 
+    protected OffsetDateTime getLastUpdateDateTime(Integer tenantId, EntityManager entityManager, Class<? extends AbstractPersistable> entityClass) {
+        final String queryString = "SELECT MAX(lastUpdateDateTime) FROM " + entityClass.getSimpleName() + " row WHERE row.tenantId = " + tenantId;
```

Comment:
Security problem, this is vulnerable to JPA-QL or SQL injection hacking.
Use setParameter("entityClass", entityClass.getSimpleName()) instead

Reply:
https://imgs.xkcd.com/comics/exploits_of_a_mom.png

Reply:
This is not exposed to client side; to perform the exploit you would have to already hacked the server, but I changed it to native query with parameter set

Reply:
Actually impossible: https://stackoverflow.com/a/31159912; I need to escape it when building the string...

Reply:
I check if the name is a valid table name, if not, I throw an IllegalStateException. Unfortunately, JPA does not provide a utility for escaping table names/verifying table names, so I have to create one from its spec